lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -2128,23 +2110,21 @@ namespace Js
         wchar_t *outStr = builder.DangerousGetWritableBuffer();
 
         wchar_t* outStrLim = outStr + count;
+        wchar_t *o = outStr;
+
+        while (o < outStrLim)
+        {
+            *o++ = *inStr++;
+        }
 
-        //TODO the mapping based on unicode tables needs review for ES5.
-        // mapping is not always 1.1
         if(toCase == ToUpper)
         {
-            while (outStr < outStrLim)
-            {
-                *outStr++ = CharToUpperCase(*inStr++);
-            }
+            CharUpperBuffW(outStr, count);
```

Comment:
Tested CharLowerW(), CharLowerBuffW(), CharUpperW(), CharUpperBuffW() and confirmed that no length increase. In other words, none of the cases in SpecialCasing.txt that involve length increase have been implemented in these APIs.

Given that these APIs (around since Windows 2000) convert strings in-place, it is unclear how they can evolve to support length increase without introducing risk of buffer overrun while maintaining backward compatibility.

Therefore I believe these APIs are good to use until a better solution (new Windows API, Unicode library etc. that provide full support of SpecialCasing.txt) becomes available.